### ISSUE TITLE
cargo returns outdated warnings

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/cargo#7862]
code:
````rust
fn main() {
    let a = 0;
}
````
compiles:
````
$cargo clippy
    Checking warn v0.1.0 (/tmp/warn)
warning: unused variable: `a`
 --> src/main.rs:2:9
  |
2 |     let a = 0;
  |         ^ help: consider prefixing with an underscore: `_a`
  |
  = note: `#[warn(unused_variables)]` on by default
````
Suppressing the warning via cli parameters does no longer work until the project is rebuilt from scratch:
````
$cargo clippy -- -Aunused_variables
warning: unused variable: `a`
 --> src/main.rs:2:9
  |
2 |     let a = 0;
  |         ^ help: consider prefixing with an underscore: `_a`
  |
  = note: `#[warn(unused_variables)]` on by default

    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
````

`cargo 1.42.0-nightly (9d32b7b01 2020-01-26)`

**Comment by aleksator** [comment]
Isn't it an issue with clippy so should be filed against that?
https://github.com/rust-lang/rust-clippy

**Comment by ehuss** [comment]
This is a known issue that will be fixed by #7533.  

You can use `cargo clippy-preview` on nightly to avoid this for now.

**Comment by aleksator** [comment]
Triage: the PR #7533 has landed, but the issue persist.

**Comment by ehuss** [comment]
To be clear, there's a second part where clippy will need to be updated to use the new functionality.  There is a rough outline [here](https://github.com/rust-lang/cargo/pull/7533#issuecomment-599133278).

**Comment by aleksator** [comment]
Seems like there is some related ongoing work [over here](https://github.com/rust-lang/rust-clippy/pull/5363).

**Comment by ehuss** [comment]
Oh, I think I misunderstood the original request here, which is specifically around changing the args passed to *clippy* (like `-Aunused_variables`).  Indeed these are not tracked in the fingerprint or hash (cargo-clippy hides them via `__CLIPPY_HACKERY__`).  

I don't offhand have ideas on how to improve that.  Probably clippy could pass those via something like `RUSTFLAGS` and Cargo could be better about including rustflags in the hash or fingerprint.

**Comment by Emilgardis** [comment]
This is quite annoying for me because I use `clippy -- --deny warnings` in a commit hook.

This also makes it difficult to use with cargo watch (used by for example Rust Analyzer) and wanting to pass `--message-format=short` or whatever you want in your own build.

**Comment by ebroto** [comment]
I think this can be closed as it was fixed in [clippy#6441](https://github.com/rust-lang/rust-clippy/pull/6441)
